[CU-f9300m] Create tournament creator view

Buglik opened the pull request:
![Zrzut ekranu z 2021-04-15 o 11 52 37](https://user-images.githubusercontent.com/9595999/115714408-5fe1c780-a377-11eb-9a29-93341b3f1fff.png)
![Zrzut ekranu z 2021-04-15 o 11 52 08](https://user-images.githubusercontent.com/9595999/115714420-640de500-a377-11eb-9984-0468a3737e70.png)
![Zrzut ekranu z 2021-04-15 o 11 51 57](https://user-images.githubusercontent.com/9595999/115714440-68d29900-a377-11eb-9f2d-f43b47858c5c.png)
![Zrzut ekranu z 2021-04-15 o 11 50 04](https://user-images.githubusercontent.com/9595999/115714469-6ec87a00-a377-11eb-91dc-14786b0488b0.png)
![Zrzut ekranu z 2021-04-15 o 11 50 08](https://user-images.githubusercontent.com/9595999/115714461-6d974d00-a377-11eb-98f9-eed7be8fc4fb.png)


White border at mobile "+CREATE" button is just a bug of the screenshot taken by firefox :)

Zummek committed:
Add create tournament endpoint [WIP]
waiting for db model

Zummek committed:
Add create tournament endpoint [WIP]
waiting for db model

Zummek committed:
Change conception for controllers

Zummek committed:
Change conception for controllers

Zummek committed:
Add writeResponse helper

Zummek committed:
Add writeResponse helper

Zummek committed:
Update routes

Zummek committed:
Update routes

Zummek committed:
Update writeResponse helper

Zummek committed:
Update writeResponse helper

Zummek committed:
Remove writeResponse helper, update routes
this project is too simple to use writeResponse helper

Zummek committed:
Remove writeResponse helper, update routes
this project is too simple to use writeResponse helper

Zummek committed:
Update tournament router - add '/' to post endpoint

Zummek committed:
Update tournament router - add '/' to post endpoint

Buglik committed:
a lot of work to go
TODO:
- getting rid of "val is null" error after adding a team
- create separate component for team builder
- add lastName to select option-lebel
- improve validation
- prepare mobile view

Buglik committed:
a lot of work to go
TODO:
- getting rid of "val is null" error after adding a team
- create separate component for team builder
- add lastName to select option-lebel
- improve validation
- prepare mobile view

Buglik committed:
changed directory

Buglik committed:
changed directory

Buglik committed:
moved teamBuilder to a component

Buglik committed:
cleaning up

Buglik committed:
added selected-item slot template

Buglik committed:
button "create" disabled when no teams added

Buglik committed:
tournament type section commented

Buglik committed:
filtering by firstName(only)
(code commented cause of multiple slint errors)

Buglik committed:
filtering by firstName(only)
(code commented cause of multiple slint errors)

Buglik committed:
added full filtering (first + lastName)
improved validation as well
got rid of select dialog behavior

Buglik committed:
added full filtering (first + lastName)
improved validation as well
got rid of select dialog behavior

Zummek commented:
Task linked: [CU-f9300m Create tournament creator view](https://app.clickup.com/t/f9300m)

eastemployeer commented on web-app/src/components/User.ts:
```diff
@@ -0,0 +1,8 @@
+  lastName: string;
+  email: string;
+  avatarSrc: string;
+}
```
What is it doing in this location?

eastemployeer commented on web-app/src/components/tournament/creator/teamBuilder.vue:
```diff
@@ -0,0 +1,214 @@
+
+  private submitAddTeam() {
+    // validation
+    if (this.teamName === '') {
+      this.isErrorTeamName = true;
+    } else {
+      this.isErrorTeamName = false;
+    }
+    if (!this.player1) {
+      this.isErrorPlayer1 = true;
+    } else {
+      this.isErrorPlayer1 = false;
+    }
+    if (this.player2 === this.player1 && this.player2) {
+      this.isErrorPlayer2 = true;
+    } else {
+      this.isErrorPlayer2 = false;
+    }
+
+    if (this.isErrorTeamName || this.isErrorPlayer1 || this.isErrorPlayer2) {
+      return;
+    }
```
Almost every if/else block from this piece of code is unnecessary

Buglik replied:
148-161 checks (and reset) "team builder" form validation
164 is there to stop from submiting form when sth not right,
i've done it like this to show all wrong inputs not only the first found.

eastemployeer replied:
I know what this code is doing........... the problem is HOW it is doing it. Low quality code

eastemployeer commented on web-app/src/components/tournament/creator/teamBuilder.vue:
```diff
@@ -0,0 +1,214 @@
+  private resetFilters() {
+    this.isErrorPlayer1 = false;
+    this.isErrorPlayer2 = false;
+    this.isErrorTeamName = false;
```
These names mean nothing to me

eastemployeer commented on web-app/src/components/tournament/creator/teamBuilder.vue:
```diff
@@ -0,0 +1,214 @@
+      this.isErrorPlayer2 = false;
+    }
+
+    if (this.isErrorTeamName || this.isErrorPlayer1 || this.isErrorPlayer2) {
+      return;
+    }
```
WHAT IS THIS?

Zummek replied:
Move/Split this to `validation()` function for more readable

eastemployeer commented on web-app/src/components/models.ts:
```diff
@@ -1,8 +1,54 @@
+  owner: User;
+  teams: Array<Team>;
+  matches: Array<Match>;
+}
```
And sharedTypes are for?

eastemployeer commented on web-app/src/components/tournament/creator/teamBuilder.vue:
```diff
@@ -0,0 +1,214 @@
+  private filterOptions: User[] = [];
+
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  private selectFilter(val: string, update: any) {
```
any spotted :) Change requested :)

Buglik replied:
done so after consultation with @Zummek ;)

Zummek replied:
This `update` it's crazy object is something like `(() => void) => void` but it doesn't work. Maybe need a few min to resolve this

Buglik replied:
added task in backlog at Clickup to resolve this later

eastemployeer commented on web-app/src/components/tournament/creator/teamsList.vue:
```diff
@@ -0,0 +1,70 @@
+import { IPagination, IColumns, Team } from 'src/components/models';
+
+@Component
+export default class teamsList extends Vue {
```
Class names begin with upper case

eastemployeer reviewed (changes requested):
Some of the files are very very long. Can't they be simplified? By splitting to several components or maybe other way? I am not even trying to understand this whole bunch of html

Zummek committed:
Lint app, update package-lock.json

Zummek committed:
Lint app, update package-lock.json

Buglik committed:
changed shared types
and put it in a good place :)

Buglik committed:
changed shared types
and put it in a good place :)

Buglik committed:
created separate validation function
changed validationBooleans names as well

Buglik committed:
created separate validation function
changed validationBooleans names as well

Buglik committed:
added maxwidth with centering and little cleanup

Zummek committed:
temporary as template

Zummek committed:
temporary as template

Zummek committed:
Add tournament-generator package/typings,

Zummek committed:
Add tournament-generator package/typings,

Zummek committed:
Try to fix cicd again

Zummek committed:
Try to fix cicd again

Zummek committed:
Try to run cicd

Zummek committed:
Try to run cicd

Zummek committed:
Lint code, and try to fix cicd :P

Zummek committed:
Lint code, and try to fix cicd :P

Zummek committed:
Update codeLinter

Zummek committed:
Update codeLinter

Zummek committed:
Add possibility to update match outcames

Zummek committed:
Add possibility to update match outcames

Zummek committed:
Update types for use '_' prefix to id property, fix getById

Zummek committed:
Update types for use '_' prefix to id property, fix getById

Buglik committed:
added locale

Buglik committed:
cleanup

Buglik committed:
imports directories changed

Buglik committed:
Cookies beeing checked (WiP)

Buglik committed:
Cookies beeing checked (WiP)

Buglik committed:
fixed types after temporary changes

Buglik committed:
fixed types after temporary changes

Buglik committed:
Added redirecting

Buglik committed:
Added redirecting

Zummek commented on web-app/src/boot/i18n.ts:
```diff
@@ -14,7 +14,10 @@ Vue.use(VueI18n);
+  messages,
+  // locale: 'pl',
+  // fallbackLocale: 'pl',
+  // messages,
```
remove comments

Zummek commented on web-app/src/components/tournament/creator/teamBuilder.vue:
```diff
@@ -0,0 +1,215 @@
+          type="submit"
+          color="primary"
+          class="q-mt-md"
+          :style="$q.screen.gt.xs ? '' : 'width:100%'"
```
If I remember correctly You can use something like this: `:style="{$q.screen.gt.xs: 'width: 100%'}"`

Zummek commented on web-app/src/components/tournament/creator/teamBuilder.vue:
```diff
@@ -0,0 +1,215 @@
+          <template v-slot:no-option>
+            <q-item>
+              <q-item-section class="text-grey">
+                No results
```
locales

Zummek commented on web-app/src/components/tournament/creator/teamBuilder.vue:
```diff
@@ -0,0 +1,215 @@
+          <template v-slot:no-option>
+            <q-item>
+              <q-item-section class="text-grey">
+                No results
```
locales

Zummek commented on web-app/src/components/tournament/creator/teamBuilder.vue:
```diff
@@ -0,0 +1,215 @@
+    } else return true;
+  }
+
+  private filterOptions: User[] = [];
```
move class variables to top of class, variables > functions

Zummek commented on web-app/src/components/tournament/creator/teamsList.vue:
```diff
@@ -0,0 +1,69 @@
+        {{ $t('tournament.noTeamsAddedErrorCreator') }}
+      </div>
+    </template>
+    <!-- <template v-slot:top-right>
+                <q-btn dense icon="add" color="primary" class="q-pr-xs">
+                  Add new team
+                </q-btn>
+              </template> -->
```
```suggestion
```
Is it useless?

Buglik replied:
yeah, it will "disappear" :P

Zummek commented on web-app/src/components/tournament/creator/teamsList.vue:
```diff
@@ -0,0 +1,69 @@
+          </q-card-section>
+          <q-card-section class="col-1 q-mr-md">
+            <q-btn dense icon="remove" @click="deleteTeam(props.row)" />
+          </q-card-section>
```
For something like this is dedicated component `q-card-action`
https://quasar.dev/vue-components/card#with-actions

Zummek commented on web-app/src/i18n/en-us/tournament.ts:
```diff
@@ -0,0 +1,15 @@
+export default {
+  name: 'Tournament Name',
+  team: 'Team',
+  teams: 'Teams List',
```
```suggestion
  teamList: 'Team List',
```
teams != "team list"
When you will want to use `teams` word there may be some problems

Zummek commented on web-app/src/i18n/en-us/tournament.ts:
```diff
@@ -0,0 +1,15 @@
+  teamName: 'Team name',
+  teamCapitan: 'Captian',
+  teamSecondMember: 'Second player',
+  teamSecondMemberCreator: 'Secobd player (optional)',
```
Hmm maybe remove team prefix? 
When it's optional add this word to key

If you want to separate subnamespace use nested object 

```
...

team: {
   "bla": "abc",
   ...
}
...
```

You can get access to this through t("tournament.team.bla")

Capitan => Captain,
Secobd => Second

Zummek commented on web-app/src/i18n/en-us/tournament.ts:
```diff
@@ -0,0 +1,15 @@
+  teamCapitan: 'Captian',
+  teamSecondMember: 'Second player',
+  teamSecondMemberCreator: 'Secobd player (optional)',
+  teamAdd: 'Add team',
```
```suggestion
  addTeam: 'Add team',
```
hmm, what do you think?

Buglik replied:
changing to team.add (as above)

Zummek commented on web-app/src/i18n/en-us/tournament.ts:
```diff
@@ -0,0 +1,15 @@
+  teamSecondMemberCreator: 'Secobd player (optional)',
+  teamAdd: 'Add team',
+  teamBuilderLabel: 'Team creator',
+  tournamentCreatorLabel: 'Tournament creator',
```
`label` postfix its good when You want to separate similar names ...Lable, ...Placeholder, ....Value

Zummek commented on web-app/src/i18n/pl/index.ts:
```diff
@@ -0,0 +1,7 @@
+export default {
+  common,
+  tournament,
+};
```
\n

Zummek commented on web-app/src/layouts/MainLayout.vue:
```diff
@@ -127,7 +127,14 @@
 <script lang="ts">
-const linksData = [
+
+
```
Redundant lines

Zummek commented on web-app/src/pages/Login.vue:
```diff
@@ -61,6 +61,8 @@ export default class Login extends Vue {
     if (this.$route.query.code) {
       await api.auth.getCookieToken(this.$route.query.code as string);
+
+      await this.$router.push({ name: 'TournamentsList' });
```
```suggestion
      await this.$router.push({ name: 'TournamentList' });
```

I'm not sure, that is best way to change routes. 
You definitely shouldn't use `push` only `replace`. It's silly to allow user to back to login page

Zummek commented on web-app/src/pages/tournament/TournamentCreator.vue:
```diff
@@ -0,0 +1,177 @@
+      <div class="row justify-between">
+        <div
+          :class="$q.screen.gt.xs ? 'col-6' : 'col-12'"
+          :style="$q.screen.gt.xs ? 'max-width: 300px' : ''"
```
try here with `:style="{...}" , but hmm i'm not sure, that is super more readable

Buglik replied:
I'll leave it as it is.
For me it's way more clean.

Zummek commented on web-app/src/pages/tournament/TournamentCreator.vue:
```diff
@@ -0,0 +1,177 @@
+    // SENDING THAT CRAP FAAAAR AWAY
+
+    // eslint-disable-next-line @typescript-eslint/no-floating-promises
+    this.$router.push({ name: 'TournamentsList' });
```
Maybe push tournament Details with id of created tournament

Zummek commented on web-app/src/router/index.ts:
```diff
@@ -25,14 +27,11 @@ export default route<Store<StateInterface>>(function({ Vue }) {
       next({
-        path: '/login',
-        query: {
-          next: to.path,
-        },
+        name: 'login',
```
Hmm? Why? That was fine. Next property points to route which redirecting you after log in [something to add in future]. If you removed this, someone adds this in future task abut this feature

Buglik replied:
gonna bring it back then

Zummek reviewed (commented):
So close, trifles

Buglik committed:
minor changes

Zummek commented on web-app/src/components/tournament/creator/teamsList.vue:
```diff
@@ -41,9 +36,9 @@
             </div>
           </q-card-section>
-          <q-card-section class="col-1 q-mr-md">
+          <q-card-actions class="col-1 q-mr-lg">
```
are classes  required?

Buglik replied:
yeah, for display purpose

Zummek commented on web-app/src/i18n/pl/common.ts:
```diff
@@ -3,4 +3,5 @@ export default {
   cancel: 'Anuluj',
   create: 'Stwórz',
   name: 'Nazwa',
+  noResults: 'Brak wyników wyszukiwania'
```
NoResults != NoSearchResults
So is this a search result or some other results?
This is a potential name for error because this is a common locale.

Zummek reviewed (approved):
LGTM, Look to two comments below
Make rebase to main :P

Buglik committed:
locales attribute changed

Buglik committed:
locales attribute changed

Buglik committed:
adding API connection

Buglik committed:
adding API connection

Buglik committed:
API connected
changed validation a bit as well

Buglik committed:
API connected
changed validation a bit as well

Buglik committed:
adding changes to layout (desktop)
DONE:
-teamsList component desktop view as normal list (not grid) with virtual scroll
- tournamentName as popup edit
- create button moved to the top-right
TODO:
-team builder component changes

Buglik committed:
some updates
teamBuilder updated
TournCreator with minor changes

Zummek commented on server/concepts/Tournament.ts:
```diff
@@ -64,7 +64,7 @@ export default class Tournament implements TournamentWihtoutMS {
     const assignedTeam = match.getAssignedTeam(currentUserId);
 
-    if (!assignedTeam) throw new AppError('You are not authorized to update this match', 403);
+    if (!assignedTeam) throw new AppError('You are not authorized to update this match', 401);
```
Why did You change this? 403 is right status

Buglik replied:
That's what came with a rebase that i was actually doing with you :P

Zummek replied:
I only watched for correct commands. You did the rebase yourself

Zummek commented on server/controllers/authController.ts:
```diff
@@ -15,7 +15,8 @@ const cookieOptions: CookieOptions = {
     Date.now() + +process.env.JWT_COOKIE_EXPIRES_IN_HOURS! * 60 * 60 * 1000,
   ),
-  httpOnly: true,
+  // TODO: temporary
```
remove com

Zummek commented on server/database/repositories/TournamentRepository.ts:
```diff
@@ -20,6 +20,13 @@ export default class TournamentRepository {
+    //     set[`matches.$.${field}`] = `${match[field as keyof Match]}`;
+    //   }
+    // });
+
```
remove

Zummek commented on web-app/src/components/tournament/creator/teamBuilder.vue:
```diff
@@ -0,0 +1,234 @@
+                  />
+                </q-item-label>
+              </q-item-section>
+            </q-item>
```
I think some optimization is needed here. Too many elements
Look to quasar doc/examples
Try to use one `q-item`, 
don't use `q-item-label` when You don't use label.

Or you may don't use `q-item` and code this with divs

Zummek commented on web-app/src/components/tournament/creator/teamBuilder.vue:
```diff
@@ -0,0 +1,234 @@
+      </q-card-section>
+      <q-card-section class="row items-stretch content-center q-pb-none">
+        <q-select
+          :class="$q.screen.gt.xs ? 'col-11 q-pb-none' : 'col-12'"
```
```suggestion
          class="col-xs-11 col-12 q-pb-xs-none"
```
If I remember correctly, it will work

Zummek commented on web-app/src/components/tournament/creator/teamBuilder.vue:
```diff
@@ -0,0 +1,234 @@
+        :style="$q.screen.gt.xs ? 'position:absolute; bottom:0' : ''"
+      >
+        <q-btn
+          :disabled="players.length < 1"
```
```suggestion
          :disabled="!players.length"
```

Zummek commented on web-app/src/components/tournament/creator/teamBuilder.vue:
```diff
@@ -0,0 +1,234 @@
+          class="q-mt-md"
+          :style="$q.screen.gt.xs ? '' : 'width:100%'"
+        >
+          <q-tooltip v-if="players.length < 1" content-class="bg-accent">
```
```suggestion
          <q-tooltip v-if="!players.length" content-class="bg-accent">
```

Zummek commented on web-app/src/components/tournament/creator/teamBuilder.vue:
```diff
@@ -0,0 +1,234 @@
+    this.$emit('team-added', teamToAdd);
+
+    this.teamName = '';
+    this.player = null;
```
If it's a temporary add-only player, can it be renamed to `tempPlayer`?

Zummek commented on web-app/src/components/tournament/creator/teamBuilder.vue:
```diff
@@ -0,0 +1,234 @@
+        const needle = val.toLowerCase();
+        this.filterOptions = this.users.filter(
+          user =>
+            // eslint-disable-next-line @typescript-eslint/restrict-plus-operands
+            (user.firstName + ' ' + user.lastName)
```
`eslint-disable-next-line` is required? 
`@typescript-eslint/restrict-plus-operands` occurs when "When adding two variables, operands must both be of type number or of type string"
firstName and lastName should be strings hmm

Buglik replied:
yeah but it popped out as an error, maybe just some VS bug

Zummek commented on web-app/src/components/tournament/creator/teamBuilder.vue:
```diff
@@ -0,0 +1,234 @@
+          :style="$q.screen.gt.xs ? '' : 'width:100%'"
+        >
+          <q-tooltip v-if="players.length < 1" content-class="bg-accent">
+            {{ $t('tournament.atLeastOneMemberError') }}
```
```suggestion
            {{ $t('tournament.team.atLeastOneMemberError') }}
```
Does this error occur only in team

Zummek commented on web-app/src/components/tournament/creator/teamsList.vue:
```diff
@@ -0,0 +1,115 @@
+    <template v-slot:no-data>
+      <div class="full-width row flex-center text-accent q-gutter-sm">
+        <q-icon size="2em" name="sentiment_dissatisfied" /><br />
+        {{ $t('tournament.noTeamsAddedErrorCreator') }}
```
Does this error occur only in team?

Zummek replied:
My point is, it can be moved to the tournament.team namespace

Zummek commented on web-app/src/components/tournament/creator/teamsList.vue:
```diff
@@ -0,0 +1,115 @@
+</script>
+<style lang="sass">
+.sticky-virtscroll-table
+  /* height or max-height is important */
```
```suggestion
```

Zummek commented on web-app/src/i18n/en-us/tournament.ts:
```diff
@@ -0,0 +1,23 @@
+    add: 'Add team',
+    builderTitle: 'Team creator',
+  },
+
```
remove empty line, team object move to end, it's more readable

Zummek commented on web-app/src/pages/tournament/TournamentCreator.vue:
```diff
@@ -0,0 +1,228 @@
+  private users: User[] = [];
+
+  private async submitAddTournament() {
+    // console.log('Creating tournament');
```
```suggestion
```

Zummek commented on web-app/src/pages/tournament/TournamentCreator.vue:
```diff
@@ -0,0 +1,228 @@
+      user.avatarSrc = 'https://cdn.quasar.dev/img/boy-avatar.png';
+    });
+  }
+  // avatarSrc: 'https://cdn.quasar.dev/img/boy-avatar.png',
```
```suggestion
```

Zummek commented on web-app/src/services/API/apiResources/types.ts:
```diff
@@ -23,12 +24,17 @@ export interface GetUserPhotoResponse {
 
-export interface CreateTournamentPayload {
-  name: string; // TODO: it is temporary as template
+// eslint-disable-next-line @typescript-eslint/no-empty-interface
+export interface CreateTournamentPayload extends tournamentCreate {}
```
try something like `export type UpdateTournamentMatchPayload = UpdateMatchOutcomes;`

Zummek reviewed (commented):
@Buglik

Buglik committed:
updates after CR

Buglik committed:
updates after CR

Buglik committed:
changes after rebase

eastemployeer committed:
Add necessary changes

eastemployeer reviewed (approved):
DONE
